### ISSUE TITLE
How to train NNUE model?

### ISSUE THREAD

**Issue by swim2sun** [thomasahle/sunfish#119]
Thank you for your work—this is an excellent project! I'm particularly interested in the NNUE version. I noticed that the code loads a pre-trained model. Could you please explain how the model was trained?

**Comment by agnelvishal** [comment]
Yes. I also would like to know how the .pickle file was generated.
The Stockfish nnue project generates .nnue file and not .pickle file.

**Comment by thomasahle** [comment]
I'm afraid I don't have the code anymore. But it was trained on cclr data: https://lczero.org/blog/2018/09/a-standard-dataset/
This data is not enhanced with evaluations, so it just uses win/loss/draw labels. Probably using mse-loss like alpha-zero.
I tried a lot of different compact architectures, as you can see in my w&b below. However, sadly I never got it fast enough to be stronger than the original sunfish :(
<img width="1518" alt="Image" src="https://github.com/user-attachments/assets/05445da3-34d8-498d-90ce-57ace637da54" />

**Comment by thomasahle** [comment]
I've written up what I remember in `nnue/README.md` on the nnue branch: https://github.com/thomasahle/sunfish/blob/nnue-mutable-board/nnue/README.md

Short version: trained on the CCRL dataset published by the lc0 project, using only win/draw/loss outcomes as labels with an MSE-style loss (like AlphaZero). The original training code is unfortunately lost. Note the NNUE version now lives on that branch until it's fast enough to be worth using.

**Comment by foo123** [comment]
@thomasahle maybe using [self-play algorithm](https://en.wikipedia.org/wiki/Self-play), appropriately adjusted, can help in training the NNUE better

**Comment by thomasahle** [comment]
Honestly I'm having a lot of trouble getting useful performance out if the 1024 byte NNUE. 
If you can make it work with self play, I'd be very interested in the results!

**Comment by foo123** [comment]
@thomasahle  my rough idea is 1) have some players with random weights play against each other multiple rounds, 2) do next stage by having new players having their weights tuned to the winners of the last stage along with some randomization and/or tuning, 3) continue until some strong NNUE arises out of the process. The idea is simple, the challenge is how to best tune and randomize the weights from one stage to the next.. 

Maybe next stage should include players each having the weights of the winner, but each with slightly changing only one or couple different weights randomly (along with the original winner).

In a sense, it would be like a combination of self-play and evolutionary programming

Eg see [Neuroevolution](https://en.wikipedia.org/wiki/Neuroevolution)
